i#725: add attach for windows

OrBenPorath opened the pull request:
Adding attach functionality on windows [i#725](https://github.com/dynamorio/dynamorio/issues/725).
Completing [this PR](https://github.com/DynamoRIO/dynamorio/pull/3328).

Main difference from the original PR, is not taking over threads that are terminating (otherwise attach always fails). 
Additionally, added the possibility to sleep 1 millisecond between takeover attempts, and controlling the number of attempts.

Feature marked as experimental.

derekbruening commented on core/CMakeLists.txt:
```diff
@@ -333,9 +333,9 @@ endif (WIN32)
 if (WIN32)
   if (DEBUG)
-    set(WIN32_C_LIB libcmtd)
+    set(WIN32_C_LIB libcmtd Psapi)
```
Please go through all my review comments on the original PR #3328 and address them (or explain in a review comment why they do not apply) and when done re-request a review.  E.g. see https://github.com/DynamoRIO/dynamorio/pull/3328/files#r246456892 which seems to have been ignored in this PR.

OrBenPorath replied:
Will do. I made the mistake of assuming all resolved conversations are resolved...

For this specific comment, psapi is actually not used anymore, so I just removed it.
Addressed most of the other comments as well.
Would love some pointers on how to address the documentation and testing parts.

derekbruening replied:
> I made the mistake of assuming all resolved conversations are resolved

Hmm, did not realize some of those are resolved: that is strange.  Not sure if the original author recently resolved and then failed to push the changes or sthg.

> Would love some pointers on how to address the documentation and testing parts.

Starting by bringing the content from the prior comments on docs: from https://github.com/DynamoRIO/dynamorio/pull/3328#discussion_r246552961 here:

+ Maybe the main documentation ("page_deploy" in intro.dox) should talk about attaching as a supported feature?
+ It should be listed in the changelog in release.dox for sure as a major feature.

And on testing https://github.com/DynamoRIO/dynamorio/pull/3328#discussion_r246553938

+ Would it be possible to write some simple tests of these new attach features? It will take a little more work than adding a regular test because you'll need to launch something in the background. See client.nudge_test and how it has a .runall file and uses run_in_bg to run the infloop app in the background.
+ If you'd prefer to make adding a test a separate CL that would be fine.

OrBenPorath replied:
>* Maybe the main documentation ("page_deploy" in intro.dox) should talk about attaching as a supported feature?
>* It should be listed in the changelog in release.dox for sure as a major feature.

Added. Is it in the correct place?

>* Would it be possible to write some simple tests of these new attach features? It will take a little more work than adding a regular test because you'll need to launch something in the background. See client.nudge_test and how it has a .runall file and uses run_in_bg to run the infloop app in the background.
>* If you'd prefer to make adding a test a separate CL that would be fine.

Can you explain the structure of the nudge_test files? How does it use run_in_bg?

derekbruening replied:
> Can you explain the structure of the nudge_test files? How does it use run_in_bg?

See https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/CMakeLists.txt#L1304 and https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/runall.cmake#L32

derekbruening replied:
> Added. Is it in the correct place?

Yes.

OrBenPorath replied:
I'm getting lost with the tests (and for some reason I am unable to run them locally).

If I understand correctly, I need to add 3 files as in `nudge_test`, and then add to `suite\tests\CMakeLists.txt` something like:
`  tobuild_ci(client.attach_test client-interface/attach_test.runall "" "-attach <pid>" "")`
but I don't think I have the process id at that point...

What am I missing? Should I add a new type of runall?

derekbruening replied:
> but I don't think I have the process id at that point...

The nudge test also needs the pid and it gets it from a file that run_in_bg writes for it: look for "pidfile".and 

> for some reason I am unable to run them locally

Run "ctest -V -R nudge_test" in a build directory where BUILD_TESTS and DEBUG where both configured as ON.

OrBenPorath replied:
I've added the test. I think it creates the target process with `run_in_bg` properly (not under DR so we can test the attach), and runs an ok `drrun` attach command.
But for some reason it is still failing. Can you take a look?

derekbruening replied:
You're saying it passes locally but not on GA CI?  Supplying more detals on the failure and pasting the relevant logs would make it easier to get help -- I went and got them myself:

```
2021-09-10T23:43:19.4669625Z         Start 108: code_api|client.attach_test
2021-09-10T23:43:19.4670098Z 
2021-09-10T23:43:19.4675119Z 108: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-D" "toolbindir=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64" "-D" "cmd=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/run_in_bg@-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/drrun.exe@-pidfile@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid@-no_inject@--@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/win32.infloop.exe" "-D" "out=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out" "-D" "nudge=<attach>@-quiet@-debug@-use_dll@D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll@-exit0@-stderr_mask@0xC@-msgbox_mask@0@-dumpcore_mask@0x7d@-staged@-code_api@-c@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.attach_test.dll.dll" "-D" "clear=" "-D" "pidfile=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid" "-P" "D:/a/dynamorio/dynamorio/suite/tests/runall.cmake"
2021-09-10T23:43:19.4680091Z 108: Test timeout computed to be: 600
2021-09-10T23:46:39.6901312Z 108: CMake Error at D:/a/dynamorio/dynamorio/suite/tests/runall.cmake:268 (message):
2021-09-10T23:46:39.6902351Z 108:   Timed out waiting for "done"
```

There's just not enough info there to know whether the attach succeeded and DR then hung, or what.  I would temporarily add more verbosity: `-stderr_mask 15` to get output on DR initializing; status messages from drrun on success of the attach; etc.  (If doing a lot of diagnostic runs maybe temporarily delete all the other `.github/workflows/ci-*.yml` files to avoid running everything else every time).  If it still passes locally but not on GA CI and verbosity is not enough you would have to resort to using tmate to log in and debug that way: https://dynamorio.org/page_test_suite.html#autotoc_md261

OrBenPorath replied:
Fixed it :)
The issue was that `win32.infloop.exe` was blocking indefinitely, and the attach failed.
I changed the target to the new (and simpler) `win32.attachee.exe` and it works.

derekbruening replied:
> The issue was that `win32.infloop.exe` was blocking indefinitely, and the attach failed.

But we want to be able to attach no matter what the target is doing.  Is this a limitation of DebugActiveProcess?  We need to document what the precise limitations are and we should file an issue on them with ideas on how to solve them.  Not being able to attach to a tiny process seems like a significant limitation.

OrBenPorath replied:
Attach failed when the main thread (to which we are injecting) was blocking.

I created a new suspended thread, and injected into it, and the attach is working for `win32.infloop.exe` as well.
I created the thread so that it would sleep indefinitely, because the assert in core/win32/syscall.c:2012 fails if the injected thread exists before the process does.

One thing I don't understand (and the reason I still cannot test with `infloop`) is closing the message box. How is the message box closed in all the tests? Who is responsible for it?

derekbruening replied:
> One thing I don't understand (and the reason I still cannot test with infloop) is closing the message box. How is the message box closed in the nudge test? Who is responsible for it?

It is closed by its window title with the closewnd utility here: https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/runall.cmake#L127

derekbruening commented:
> OrBenPorath force-pushed the OrBenPorath:add-windows-attach branch 11 times

Please do not force push: it ruins the history, strands the review comments, etc.  See https://dynamorio.org/page_code_reviews.html#sec_code_review_non_member and https://dynamorio.org/page_code_reviews.html#autotoc_md114.  The final merge will squash it all down into one clean commit.

derekbruening commented on core/win32/os.c:
```diff
@@ -2497,6 +2497,15 @@ static bool
+    if (DYNAMO_OPTION(skip_terminating_threads)) {
+        if (nt_is_thread_terminating(hthread)) {
+            // takeover fails when attaching and trying to takeover terminating threads
+            // luckily, we don't really need to take over them
```
style nit: Use capitalized first words and end punctuation for multi-sentence comments (some old code does not have this new style convention but we prefer it for all new code)

OrBenPorath replied:
Done

derekbruening commented on core/win32/os.c:
```diff
@@ -2497,6 +2497,15 @@ static bool
     bool success = true;
+
+    if (DYNAMO_OPTION(skip_terminating_threads)) {
+        if (nt_is_thread_terminating(hthread)) {
```
I wonder if we can avoid the extra call to NtSuspendThread here by having nt_thread_suspend return the NTSTATUS -- but it gets more complicated if we want to support `suspended==true` so maybe simplest to leave it how you have it.

OrBenPorath replied:
My logic was "break as little as possible", but if you think we should, then I'll take a look at it.

derekbruening replied:
Just leave it.

derekbruening commented on tools/drdeploy.c:
```diff
@@ -309,6 +309,19 @@ const char *options_list_str =
+    "                          Attach to the process with the given pid.\n"
+    "                          If attach to a process which is in middle of blocking\n"
+    "                          system call, attach could fail.\n"
+    "                          try takeover_sleep and larger takeovers to increase\n"
```
nit: s/try/Try/

OrBenPorath replied:
Done

derekbruening reviewed (commented):
Thank you for the PR.  It looks pretty good: just some nits, and then the documentation and testing.  Is your plan to add a test as a separate PR?  And include some doc additions in this PR?

OrBenPorath commented on core/dynamo.c:
```diff
@@ -2912,7 +2911,9 @@ dynamorio_take_over_threads(dcontext_t *dcontext)
             bb_lock_start = true;
-    } while (found_threads && attempts < MAX_TAKE_OVER_ATTEMPTS);
+        if (DYNAMO_OPTION(sleep_between_takeovers))
+            dr_sleep(1);
```
Done

OrBenPorath commented on core/win32/injector.c:
```diff
@@ -843,6 +844,81 @@ dr_inject_process_create(const char *app_name, const char **argv, void **data OU
+        DWORD exe_path_size = MAX_PATH;
+
+        HANDLE process_handle = OpenProcess(PROCESS_ALL_ACCESS, false, (DWORD)pid);
+        if (process_handle != NULL) {
```
Done

OrBenPorath commented:
> Thank you for the PR. It looks pretty good: just some nits, and then the documentation and testing. Is your plan to add a test as a separate PR? And include some doc additions in this PR?

My plan is to finish everything in this PR.

OrBenPorath commented on api/docs/deployment.dox:
```diff
@@ -214,6 +214,12 @@ bbsize sample client using the following configure-and-run command:
+Alternatively, you can first run \c notepad.exe, and then use \c drrun
+to attach to it:
+\code
+bin32/drrun.exe -late -attach <notepad_pid> -c samples/bin32/bbsize.dll
```
Done

OrBenPorath commented on api/docs/deployment.dox:
```diff
@@ -214,6 +214,12 @@ bbsize sample client using the following configure-and-run command:
+to attach to it:
+\code
+bin32/drrun.exe -late -attach <notepad_pid> -c samples/bin32/bbsize.dll
+\endcode
```
Done

derekbruening commented:
add to whitelist

derekbruening commented:
run arm tests

OrBenPorath commented:
> > OrBenPorath force-pushed the OrBenPorath:add-windows-attach branch 11 times
> 
> Please do not force push: it ruins the history, strands the review comments, etc. See https://dynamorio.org/page_code_reviews.html#sec_code_review_non_member and https://dynamorio.org/page_code_reviews.html#autotoc_md114. The final merge will squash it all down into one clean commit.

I merged the master into this branch instead of rebasing to not force push. But I think maybe it would be better if I clean it (or just create a clean PR and copy all conversations to it).
What should I do?

derekbruening commented:
> > > OrBenPorath force-pushed the OrBenPorath:add-windows-attach branch 11 times
> > 
> > 
> > Please do not force push: it ruins the history, strands the review comments, etc. See https://dynamorio.org/page_code_reviews.html#sec_code_review_non_member and https://dynamorio.org/page_code_reviews.html#autotoc_md114. The final merge will squash it all down into one clean commit.
> 
> I merged the master into this branch instead of rebasing to not force push. But I think maybe it would be better if I clean it (or just create a clean PR and copy all conversations to it).
> What should I do?

No, please leave it.  The reviewer wants to be able to easily view just the changes since the last review; we want the full comment history; we don't want comments stranded w/o any file anchor.  The final commit message gets cleaned up on the squash-and-merge at the end.

derekbruening commented on core/lib/dr_inject.h:
```diff
@@ -112,10 +112,11 @@ DR_EXPORT
  * \param[out]  data           An opaque pointer that should be passed to
  *                             subsequent dr_inject_* routines to refer to
  *                             this process.
+ * \param[out]  app_name       pointer to the name of the target process
```
style: Capitalize and punctuate user-facing documentation sentences.

derekbruening replied:
This should point out that this pointer is only valid until dr_inject_process_exit.

OrBenPorath replied:
Done

derekbruening reviewed (commented):
Minor comments.  I think the only thing missing is understanding why we can't attach to a process with a thread that is in a wait state: does that mean we can't attach to any GUI process?

OrBenPorath commented:
> Minor comments. I think the only thing missing is understanding why we can't attach to a process with a thread that is in a wait state: does that mean we can't attach to any GUI process?

See my reply above - we only fail if the thread we are trying to inject to is blocking. I think using the new thread for injections should fixed it.

OrBenPorath commented on suite/tests/runall.cmake:
```diff
@@ -168,15 +168,17 @@ endwhile ()
-    message(FATAL_ERROR "Timed out waiting for newline")
-  endif ()
-endwhile()
+if (NOT "${nudge}" MATCHES "<attach>")
```
There is a newline, but it is only written after we attach and load the client.

derekbruening replied:
Hmm, so for the nudge test the infloop app itself doesn't print something up front and it's the client who prints it.  So adding a print to infloop might perturb the other tests.  OK.

OrBenPorath commented on suite/tests/win32/attachee.c:
```diff
@@ -0,0 +1,44 @@
+main(int argc, const char *argv[])
+{
+    for (int i = 0; i < 60; i++) {
+        Sleep(1000);
```
Done.
But currently the test always waits until the 3 minutes are done.
After I find how to close the message box it will exit immediately.

derekbruening commented on suite/tests/client-interface/attach_test.dll.c:
```diff
@@ -32,7 +32,7 @@
 #include "dr_api.h"
 
-static client_id_t client_id = 0;
+static client_id_t client_id;
```
I meant this variable is not used so it can be deleted entirely.

OrBenPorath replied:
Done

derekbruening commented on suite/tests/win32/attachee.c:
```diff
@@ -36,9 +36,7 @@
-    for (int i = 0; i < 60; i++) {
-        Sleep(1000);
-    }
+    Sleep(3 * 60 * 1000);
```
With it now working on infloop is the plan to delete this file?

OrBenPorath replied:
yes

derekbruening commented on core/win32/injector.c:
```diff
@@ -878,10 +878,17 @@ dr_inject_process_attach(process_id_t pid, void **data OUT, char **app_name OUT)
+
+    info->pi.hThread = CreateRemoteThread(
+        process_handle, NULL, 0, (LPTHREAD_START_ROUTINE)Sleep, (LPVOID)(SIZE_T)INFINITE,
+        CREATE_SUSPENDED, &info->pi.dwThreadId);
```
So this thread is present for the entire execution under DR?  That is not ideal as the app will see it if it iterates its threads (plus it called DLL_THREAD_ATTACH at creation but that's hard to avoid -- I wonder if there's a debug API solution to set the PC of a blocked thread such that we don't need this new thread).  OK to not fix for now but at least put a XXX comment about exiting the thread right after we attach for better transparency.

OrBenPorath replied:
Done. 
Is the phrasing OK?

derekbruening replied:
LGTM

OrBenPorath commented:
> > > > OrBenPorath force-pushed the OrBenPorath:add-windows-attach branch 11 times
> > > 
> > > 
> > > Please do not force push: it ruins the history, strands the review comments, etc. See https://dynamorio.org/page_code_reviews.html#sec_code_review_non_member and https://dynamorio.org/page_code_reviews.html#autotoc_md114. The final merge will squash it all down into one clean commit.
> > 
> > 
> > I merged the master into this branch instead of rebasing to not force push. But I think maybe it would be better if I clean it (or just create a clean PR and copy all conversations to it).
> > What should I do?
> 
> No, please leave it. The reviewer wants to be able to easily view just the changes since the last review; we want the full comment history; we don't want comments stranded w/o any file anchor. The final commit message gets cleaned up on the squash-and-merge at the end.

So what should I do with the new conflict?

derekbruening commented:
> So what should I do with the new conflict?

You merged master in?  Resolve like any merge.  If you clicked the web button here to "update" I think they now have a web UI to resolve it (never used it though); if you did it from the cmdline you would use your editor or whatever method you normally use.

OrBenPorath commented on suite/tests/client-interface/attach_test.dll.c:
```diff
@@ -0,0 +1,50 @@
+
+#include "dr_api.h"
+
+static client_id_t client_id = 0;
```
Done

OrBenPorath commented on suite/tests/client-interface/attach_test.dll.c:
```diff
@@ -0,0 +1,50 @@
+void
+dr_init(client_id_t id)
+{
+    client_id = id;
```
Done

derekbruening commented on suite/tests/runall.cmake:
```diff
@@ -168,7 +168,9 @@ endwhile ()
 set(iters 0)
-if (NOT "${nudge}" MATCHES "<attach>")
+if ("${nudge}" MATCHES "<attach>")
+  do_sleep(0.5)
```
If we need a real synchronization point, better to look for a sign like the loop below than just waiting and hoping the timing works out which is a recipe for flakiness.

OrBenPorath replied:
I actually don't think we need it any synchronization at this point - we only need to know that the process is up, and we verify that using the pid file.

I am just trying to understand why the test passes on my machine, but fails in the CI.

derekbruening replied:
> I think I fixed it. The issue was that because we do not verify the first line, when we verify "more input" it would be the first line, and its possible that the exit event was not registered yet, so we would timeout waiting for "done".

What was your fix for this?  The code seems to still not wait for a first line.

derekbruening replied:
Or is this the "init thread" wait?

derekbruening replied:
The output file now has the "Thank you for testing..." line so could this first line wait code be re-instated for attach?  (Maybe I"m mis-remembering what was there before...was that there before?)

derekbruening reviewed (approved):
Thank you for working through the issues of adding a test.

derekbruening commented:
Another option for debugging on the GA CI VMs is tmate: https://dynamorio.org/page_test_suite.html#autotoc_md261

OrBenPorath commented:
> Another option for debugging on the GA CI VMs is tmate: https://dynamorio.org/page_test_suite.html#autotoc_md261

That was very helpful! (and easier than I thought)

I think I fixed it. The issue was that because we do not verify the first line, when we verify "more input" it would be the first line, and its possible that the exit event was not registered yet, so we would timeout waiting for "done".
The test was unstable before that change (which is consistent with a race), so maybe its best to run it a few times before merging (is there a way to do that without pushing?).

I also moved from assuming `Sleep` is at the same address across processes to finding it dynamically. I'm not sure if that was an issue or not, but I think it is better to not make that assumption anyway.
To do so, I generalized `find_remote_ntdll_base` to `find_remote_dll_base`: it now accepts a DLL name as a parameter and looks for it. It also compares with `_strcmpi` because `kernel32` was loaded upper case. 
When I used `strcmp` and did not find anything, `find_remote_dll_base` would go into an infinite loop - I did not check the root cause of it. It does not cause any issues now, because it is only used with `ntdll` and `kernel32` which are always loaded. But now that it accepts general DLLs, I think it should be fixed.

OrBenPorath commented on api/docs/deployment.dox:
```diff
@@ -214,6 +214,12 @@ bbsize sample client using the following configure-and-run command:
+\code
+bin32/drrun.exe -late -attach <notepad_pid> -c samples/bin32/bbsize.dll
+\endcode
+
```
Done

OrBenPorath commented:
> > Another option for debugging on the GA CI VMs is tmate: https://dynamorio.org/page_test_suite.html#autotoc_md261
> 
> That was very helpful! (and easier than I thought)
> 
> I think I fixed it. The issue was that because we do not verify the first line, when we verify "more input" it would be the first line, and its possible that the exit event was not registered yet, so we would timeout waiting for "done".
> The test was unstable before that change (which is consistent with a race), so maybe its best to run it a few times before merging (is there a way to do that without pushing?).
> 
> I also moved from assuming `Sleep` is at the same address across processes to finding it dynamically. I'm not sure if that was an issue or not, but I think it is better to not make that assumption anyway.
> To do so, I generalized `find_remote_ntdll_base` to `find_remote_dll_base`: it now accepts a DLL name as a parameter and looks for it. It also compares with `_strcmpi` because `kernel32` was loaded upper case.
> When I used `strcmp` and did not find anything, `find_remote_dll_base` would go into an infinite loop - I did not check the root cause of it. It does not cause any issues now, because it is only used with `ntdll` and `kernel32` which are always loaded. But now that it accepts general DLLs, I think it should be fixed.

The test is still unstable, I will check if I missed anything else.

OrBenPorath commented:
I'm a bit stuck. Maybe you can think of something...

In most runs everything works. But in about 1/3 of the cases the test fails.

Below are the outputs from the `tmate` session:
1. First I ran the commands manually - `run_in_bg(infloop)`, `run_in_bg(drrun -attach)`, `closewnd`. Everything works as expected.
2. Next I ran `ctest -V -R attach_test`. The "done" print is missing (and therefore we are timing out).

```
runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/run_in_bg -out D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-out D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/drrun.exe -pidfile D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-pid -no_inject -- D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/bin/win32.infloop.exe

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/drview -exe win32.infloop.exePID 2912, 
Process win32.infloop.exe, running native

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/run_in_bg D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/drrun -attach 2912 -quiet -debug -use_dll D:/a/dynamorio/dynamorio/build_debug-internal-32/lib32/debug/dynamorio.dll -exit0 -stderr_mask 0xC -msgbox_mask 0 -dumpcore_mask 0x7d -staged -code_api -c D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/bin/client.attach_test.dll.dll

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/drview -exe win32.infloop.exePID 2912, 
Process win32.infloop.exe, running DR debug (build 0)

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/closewnd.exe "Infloop pid=2912" 10
Close message sent.

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# cat D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-out
thank you for testing attach
MessageBox closed
done

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# ctest -V -R attach_test
UpdateCTestConfiguration  from :D:/a/dynamorio/dynamorio/build_debug-internal-32/DartConfiguration.tcl
Parse Config file:D:/a/dynamorio/dynamorio/build_debug-internal-32/DartConfiguration.tcl
UpdateCTestConfiguration  from :D:/a/dynamorio/dynamorio/build_debug-internal-32/DartConfiguration.tcl
Parse Config file:D:/a/dynamorio/dynamorio/build_debug-internal-32/DartConfiguration.tcl
Test project D:/a/dynamorio/dynamorio/build_debug-internal-32
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 118
    Start 118: code_api|client.attach_test

118: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-D" "toolbindir=D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32" "-D" "cmd=D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/run_in_bg@-out@D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-out@D:/a/dynamorio/dynamorio/build_debug-internal-32/bin32/drrun.exe@-pidfile@D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-pid@-no_inject@--@D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/bin/win32.infloop.exe" "-D" "out=D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-out" "-D" "nudge=<attach>@-quiet@-debug@-use_dll@D:/a/dynamorio/dynamorio/build_debug-internal-32/lib32/debug/dynamorio.dll@-exit0@-stderr_mask@0xC@-msgbox_mask@0@-dumpcore_mask@0x7d@-staged@-code_api@-c@D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/bin/client.attach_test.dll.dll" "-D" "clear=" "-D" "pidfile=D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-pid" "-P" "D:/a/dynamorio/dynamorio/suite/tests/runall.cmake"
118: Test timeout computed to be: 1500
118: thank you for testing attach
118: MessageBox closed
118: 
118: CMake Error at D:/a/dynamorio/dynamorio/suite/tests/runall.cmake:272 (message):
118:   Timed out waiting for "done"
118: 
118: 
?/1 Test #118: code_api|client.attach_test ......***Failed  Required regular expression not found. Regex=[^thank you for testing attach
?essageBox closed
?one
$
]  8.89 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   8.95 sec

The following tests FAILED:
        118 - code_api|client.attach_test (Failed)
Errors while running CTest
Output from these tests are in: D:/a/dynamorio/dynamorio/build_debug-internal-32/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.

runneradmin@fv-az109-888 MINGW64 /d/a/dynamorio/dynamorio/build_debug-internal-32
# cat D:/a/dynamorio/dynamorio/build_debug-internal-32/suite/tests/client.attach_test-out
thank you for testing attach
MessageBox closed
```

Obviously there is some timing issue here. But I can't seem to find the issue.

What do you think?

derekbruening commented on core/win32/module_shared.c:
```diff
@@ -1326,6 +1326,41 @@ read_remote_maybe64(HANDLE process, uint64 addr, size_t bufsz, void *buf)
+                print_file(STDERR, "found |%s| @ 0x%I64x 64=%d\n", name, mbi.BaseAddress,
+                           is_64);
+#    endif
+                if (_strcmpi(name, dll_name) == 0 && BOOLS_MATCH(find64bit, is_64))
```
> It also compares with _strcmpi because kernel32 was loaded upper case.
>  When I used strcmp and did not find anything, find_remote_dll_base would go into an infinite loop - I did not check the root cause of it.

You mean the outer loop never terminated?  I'm not seeing how that could happen -- mbi.RegionSize being incorrect??  But we have many other loops like this.

> It does not cause any issues now, because it is only used with ntdll and kernel32 which are always loaded. But now that it accepts general DLLs, I think it should be fixed.

No kernel32 is not always loaded: early on in the life of a process there's only ntdll, and DR's earliest injection takes over well before kernel32 is loaded, so you could imagine this function being called then by DR code.  (And I think it's possible to create a process that never uses kernel32 but a pathological case for sure.)

But maybe not worth effort at this point

OrBenPorath replied:
>You mean the outer loop never terminated? I'm not seeing how that could happen -- mbi.RegionSize being incorrect?? But we have many other loops like this.

Output was something like this:

```
...
found |wow64.dll| @ 0x7ffca0d40000 64=1
0x7ffca0d41000-0x7ffca0d7a000 type=0x1000000 state=0x1000
0x7ffca0d7a000-0x7ffca0d8f000 type=0x1000000 state=0x1000
0x7ffca0d8f000-0x7ffca0d91000 type=0x1000000 state=0x1000
0x7ffca0d91000-0x7ffca0d99000 type=0x1000000 state=0x1000
0x7ffca0d99000-0x7ffca1ad0000 type=0x0 state=0x10000
0x7ffca1ad0000-0x7ffca1ad1000 type=0x1000000 state=0x1000
found |ntdll.dll| @ 0x7ffca1ad0000 64=1
0x7ffca1ad1000-0x7ffca1bec000 type=0x1000000 state=0x1000
0x7ffca1bec000-0x7ffca1c34000 type=0x1000000 state=0x1000
0x7ffca1c34000-0x7ffca1c35000 type=0x1000000 state=0x1000
0x7ffca1c35000-0x7ffca1c37000 type=0x1000000 state=0x1000
0x7ffca1c37000-0x7ffca1c40000 type=0x1000000 state=0x1000
0x7ffca1c40000-0x7ffca1cc5000 type=0x1000000 state=0x1000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
0x7ffca1cc5000-0x7fffffff0000 type=0x0 state=0x10000
...
...
...
```

>No kernel32 is not always loaded: early on in the life of a process there's only ntdll, and DR's earliest injection takes over well before kernel32 is loaded, so you could imagine this function being called then by DR code. (And I think it's possible to create a process that never uses kernel32 but a pathological case for sure.)
>
>But maybe not worth effort at this point

I see. But for now we only use kernel32 for attach, so maybe it is ok to assume it is loaded? (though clearly not the best)

derekbruening replied:
I would add a XXX or TODO or even FIXME comment to that loop explaining the infinite loop you saw.

OrBenPorath replied:
I'll do that.

In the meanwhile, I added a print of the address:
```
...
found |ntdll.dll| @ 0x7ffca1ad0000 64=1
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1ad1000
0x7ffca1ad1000-0x7ffca1bec000 size=0x11b000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1bec000
0x7ffca1bec000-0x7ffca1c34000 size=0x48000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c34000
0x7ffca1c34000-0x7ffca1c35000 size=0x1000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c35000
0x7ffca1c35000-0x7ffca1c37000 size=0x2000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c37000
0x7ffca1c37000-0x7ffca1c40000 size=0x9000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1c40000
0x7ffca1c40000-0x7ffca1cc5000 size=0x85000 type=0x1000000 state=0x1000
calling remote_query_virtual_memory_maybe64, addr=0x7ffca1cc5000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x7fffffff0000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x80035e31b000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x8006bc646000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
calling remote_query_virtual_memory_maybe64, addr=0x800a1a971000
0x7ffca1cc5000-0x7fffffff0000 size=0x35e32b000 type=0x0 state=0x10000
...
```

So something weird happens when the address passes 0x800000000000. 
Maybe just add
```
if (mbi.BaseAddress + mbi.RegionSize < addr)
    break;
```
WDYT?

derekbruening replied:
Is this infinite loop only happening for 32-bit (WOW64)?  The drrun and the target are the same bitwidth?

OrBenPorath replied:
Yes, I only tested on a 64 bit machine: it only happens for 32-bit (both the drrun and infloop).

OrBenPorath replied:
Done

derekbruening replied:
I do remember some weird Windows kernel behavior with memory query walks but not quite this one. 
 The one I'm thinking of is the AllocationBase: https://github.com/DynamoRIO/drmemory/issues/2328#issuecomment-739454412

derekbruening commented:
> Obviously there is some timing issue here. But I can't seem to find the issue.

So DR attached, the closewnd thinks it successfully closed the app (in a "nice" way to let the process run cleanup code), but DR never called the client exit callback?  I would try to repro with DR logging turned on (hopefully doesn't change timing and make failure not reproduce): what is DR doing -- does it run exit code?  Any error msgs in the logs?

derekbruening commented:
Maybe disable the other workflows again while testing Windows?

OrBenPorath commented:
> Maybe disable the other workflows again while testing Windows?

How do I do that? just remove the other yml files? or their content?

derekbruening commented:
> > Maybe disable the other workflows again while testing Windows?
> 
> How do I do that? just remove the other yml files? or their content?

Just temporarily delete the other .github/workflows/*.yml files

OrBenPorath commented on suite/tests/runall.cmake:
```diff
@@ -103,7 +103,7 @@ function (do_sleep ms)
     # XXX: ping's units are secs.  For now we always do 1 sec.
     # We could try to use cygwin bash or perl.
-    execute_process(COMMAND ${PING} 127.0.0.1 -n 1 OUTPUT_QUIET)
+    execute_process(COMMAND ${PING} 127.0.0.1 -n 2 OUTPUT_QUIET)
```
It seems that the previous command only slept a few milliseconds.

Running the following batch, the first 5 pings only take a few milliseconds, while the last takes just over 5 seconds.
```
@echo %time%
ping 127.0.0.1 -n 1
ping 127.0.0.1 -n 1
ping 127.0.0.1 -n 1
ping 127.0.0.1 -n 1
ping 127.0.0.1 -n 1
@echo %time%
ping 127.0.0.1 -n 6
@echo %time%
pause
```

derekbruening replied:
Presumably the behavior changed across Windows versions.

OrBenPorath commented:
So I think I fixed the main issue:
I moved back to the `attachee`, because I want the target to announce when it up and running, and ready to be attached.
I also changed the `do_sleep` function, because I think it was sleeping only a few milliseconds and not a full second as the remark says...

I sometimes get a weird (unrelated) error in the 64-bit test:
```
test 103
        Start 103: code_api|client.attach_test

103: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-D" "toolbindir=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64" "-D" "cmd=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/run_in_bg@-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/drrun.exe@-pidfile@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid@-no_inject@--@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/win32.attachee.exe" "-D" "out=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out" "-D" "nudge=<attach>@-quiet@-debug@-use_dll@D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll@-exit0@-stderr_mask@0xC@-msgbox_mask@0@-dumpcore_mask@0x7d@-staged@-code_api@-c@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.attach_test.dll.dll" "-D" "clear=" "-D" "pidfile=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid" "-P" "D:/a/dynamorio/dynamorio/suite/tests/runall.cmake"
103: Test timeout computed to be: 600
103: CMake Error at D:/a/dynamorio/dynamorio/suite/tests/runall.cmake:156 (string):
103:   string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
103:   command.
103: 
103: 
```

Somehow `string(REGEX REPLACE "\n" "" pid ${pid})` does not  see 6 arguments. I don't know how that could happen - maybe `${pid}` is null?

derekbruening commented:
> So I think I fixed the main issue:
> I moved back to the `attachee`, because I want the target to announce when it is up and running, and ready to be attached.
> I also changed the `do_sleep` function, because I think it was sleeping only a few milliseconds and not a full second as the remark says...
> 
> I sometimes get a weird error in the 64-bit test:
> 
> ```
> test 103
>         Start 103: code_api|client.attach_test
> 
> 103: Test command: "C:\Program Files\CMake\bin\cmake.exe" "-D" "toolbindir=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64" "-D" "cmd=D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/run_in_bg@-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out@D:/a/dynamorio/dynamorio/build_debug-internal-64/bin64/drrun.exe@-pidfile@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid@-no_inject@--@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/win32.attachee.exe" "-D" "out=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-out" "-D" "nudge=<attach>@-quiet@-debug@-use_dll@D:/a/dynamorio/dynamorio/build_debug-internal-64/lib64/debug/dynamorio.dll@-exit0@-stderr_mask@0xC@-msgbox_mask@0@-dumpcore_mask@0x7d@-staged@-code_api@-c@D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/bin/client.attach_test.dll.dll" "-D" "clear=" "-D" "pidfile=D:/a/dynamorio/dynamorio/build_debug-internal-64/suite/tests/client.attach_test-pid" "-P" "D:/a/dynamorio/dynamorio/suite/tests/runall.cmake"
> 103: Test timeout computed to be: 600
> 103: CMake Error at D:/a/dynamorio/dynamorio/suite/tests/runall.cmake:156 (string):
> 103:   string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
> 103:   command.
> 103: 
> 103: 
> ```
> 
> Somehow `string(REGEX REPLACE "\n" "" pid ${pid})` does not see 6 arguments. I don't know how that could happen - maybe `${pid}` is null?

Yes `${pid}` must be empty

derekbruening commented on core/win32/module_shared.c:
```diff
@@ -1334,6 +1334,10 @@ find_remote_dll_base(HANDLE phandle, bool find64bit, char *dll_name)
+    /* FIXME - for 32-bit targets on a 64 bit machine, this loop doesn't return,
+     * if the dll is not loaded.
+     * When addr passes 0x800000000000 remote_query_virtual_memory_maybe64
+     * returns the previous mbi (ending at 0x7FFFFFFF0000) */
```
Maybe it is best to put in the check you suggested to abort the loop (w/ this comment there as info or w/ XXX instead of FIXME): that seems the safest.

OrBenPorath replied:
Done

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -1308,9 +1308,13 @@ function(torun test key source native standalone_dr dr_ops exe_ops added_out)
-      # than to run notepad or sthg.
-      get_target_path_for_execution(app_path win32.infloop "${location_suffix}")
+      if ("${runall}" MATCHES "<attach>")
+        get_target_path_for_execution(app_path win32.attachee "${location_suffix}")
```
Add a comment explaining why attach isn't using infloop since it seems like infloop would be perfect for it

OrBenPorath replied:
Done

OrBenPorath commented:
Ok, so the test is more stable now (I am not seeing the previous errors), but still fails sometimes.

I saw two errors:
1. [Nothing happens until the test times out](https://github.com/DynamoRIO/dynamorio/runs/3685679149?check_suite_focus=true).
2. [Test output missing 'MessageBox closed' before 'done'](https://github.com/DynamoRIO/dynamorio/pull/5075/checks?check_run_id=3684533264):
```
118: starting attachee
118: thank you for testing attach
118: thread init
118: done
```

I don't understand these errors, but here are my thoughts:
1. One of the first things `attachee` does is to write to the output file (that's the whole point), and nothing else is supposed to happen until it does, so how can it be empty?
- Maybe the timeout is now less than the loops timeout, and we are not seeing anything because the process does not start, hence the pidfile is empty, and it is a different manifestation of the previous error.
2. How is the target process terminated? 
- The run takes ~12 seconds. and does not output an error before comparing the outputs, so this is not a `kill_background_process(ON)` call.
- The process does not output 'timed out' or 'MessageBox closed', so none of its threads initiated the termination.
- Maybe the injection thread somehow terminated the process? Maybe it had an exception? I added prints on exception to have some more visibility if that is the case.


Unfortunately, I will not be available to work on this next week (and probably less available afterwards as well).
How do you want to continue?

OrBenPorath commented:
> 2. How is the target process terminated?
> 
> * The run takes ~12 seconds. and does not output an error before comparing the outputs, so this is not a `kill_background_process(ON)` call.
> * The process does not output 'timed out' or 'MessageBox closed', so none of its threads initiated the termination.
> * Maybe the injection thread somehow terminated the process? Maybe it had an exception? I added prints on exception to have some more visibility if that is the case.

But now [it happened again](https://github.com/DynamoRIO/dynamorio/pull/5075/checks?check_run_id=3689664637), and there were no prints from the exception function.

What am I missing?

derekbruening commented:
> Unfortunately, I will not be available to work on this next week (and probably less available afterwards as well).
> How do you want to continue?

One way forward is to add this new test to the flaky ignore list in suite/runsuite_wrapper.pl (find the lists for the appropriate Windows bitwidth(s); put #725 in the comment), update #725 to say that the test still needs work and to be removed from the flaky list (o/w nobody will notice if attach gets broken and the test fails every time), and merge this in.

derekbruening commented:
add to whitelist

derekbruening commented:
run arm tests

OrBenPorath commented:
> One way forward is to add this new test to the flaky ignore list in suite/runsuite_wrapper.pl (find the lists for the appropriate Windows bitwidth(s); put #725 in the comment), update #725 to say that the test still needs work and to be removed from the flaky list (o/w nobody will notice if attach gets broken and the test fails every time), and merge this in.

Added to flaky list - is it in the correct place?

Should I just comment on #725 something like 
"Added in #5075.
Test still requires works, and to be removed from the flaky list."?

derekbruening commented:
> > One way forward is to add this new test to the flaky ignore list in suite/runsuite_wrapper.pl (find the lists for the appropriate Windows bitwidth(s); put #725 in the comment), update #725 to say that the test still needs work and to be removed from the flaky list (o/w nobody will notice if attach gets broken and the test fails every time), and merge this in.
> 
> Added to flaky list - is it in the correct place?

Yes, LGTM.  (Huh the existing list for Windows is longer than I thought...need more devs to help clean it up.)

> Should I just comment on #725 something like
> "Added in #5075.
> Test still requires works, and to be removed from the flaky list."?

SGTM, maybe pasting in the details of the failures you had above.

derekbruening commented:
Merged.  Thank you for the PR, and working through all the debugging to get the test as far as it is!

OrBenPorath commented:
> Merged. Thank you for the PR, and working through all the debugging to get the test as far as it is!

My pleasure :)
And thank you for DynamoRIO, it is amazing, and an honor to contribute to.
